modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/passthru/connections/HostConnections.java

```diff
@@ -63,27 +96,64 @@ public HostConnections(HttpRoute route, int maxSize) {
     public NHttpClientConnection getConnection() {
         lock.lock();
         try {
-            if (freeConnections.size() > 0) {
+            while (!freeConnections.isEmpty()) {
                 if (log.isDebugEnabled()) {
                     log.debug("Returning an existing free connection " + route);
                 }
                 NHttpClientConnection conn = freeConnections.get(0);
-                freeConnections.remove(conn);
-                busyConnections.add(conn);
-                return conn;
+                long currentTime = System.currentTimeMillis();
+                long connectionInitTime = (Long) conn.getContext().getAttribute(PassThroughConstants.
+                                                                                        CONNECTION_INIT_TIME);
+                long lastReleasedTime = (Long) conn.getContext().getAttribute(PassThroughConstants.
+                                                                                      CONNECTION_RELEASE_TIME);
+                if (isMaximumLifeSpanExceeded(currentTime, connectionInitTime) || isIdleTimeExceeded(currentTime,
+                                                                                                     lastReleasedTime)) {
+                    freeConnections.remove(conn);
+                    try {
+                        conn.shutdown();
+                    } catch (IOException io) {
+                        log.error("Error occurred while shutting down connection." + io.getMessage());
```

Comment:
Let's log the trowable as well

Reply:
Fixed in https://github.com/wso2/wso2-synapse/pull/1236